### ISSUE TITLE
[expo-firebase-analytics] Firebase analytics network requests failing

### ISSUE THREAD

**Issue by valdrox** [expo/expo#10457]
### Summary of Issue 
I've been trying to upgrade from SDK 38 to 39 and have been encountering a lot of warnings on the Android client that I don't understand. I would ignore them, but it seems related to other parts of my code crashing (the "warning" would be an "error" with the same message). 

### Environment - output of `expo diagnostics` & the platform(s) you're targeting
  Expo CLI 3.27.12 environment info:
    System:
      OS: Linux 5.4 Ubuntu 20.04.1 LTS (Focal Fossa)
      Shell: 5.0.17 - /bin/bash
    Binaries:
      Node: 14.12.0 - ~/.nvm/versions/node/v14.12.0/bin/node
      Yarn: 1.22.4 - /usr/bin/yarn
      npm: 6.14.8 - ~/.nvm/versions/node/v14.12.0/bin/npm
    npmPackages:
      expo: ~39.0.2 => 39.0.3 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-39.0.2.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.14 
    npmGlobalPackages:
      expo-cli: 3.27.12
    Expo Workflow: managed

targeting : managed expo on android and ios.

### Reproducible Demo
package.json (default)
```
"dependencies": {
    "expo": "~39.0.2",
    "expo-status-bar": "~1.0.2",
    "react": "16.13.1",
    "react-dom": "16.13.1",
    "react-native": "https://github.com/expo/react-native/archive/sdk-39.0.2.tar.gz",
    "react-native-web": "~0.13.12",
    "expo-firebase-analytics": "~2.5.0",
    "firebase": "7.9.0"
  },
 "devDependencies": {
    "@babel/core": "~7.9.0"
  },
```
babel.config.js (default, although I suspect it has to do with lazyImports)
```
module.exports = function(api) {
  api.cache(true);
  return {
    presets: ['babel-preset-expo'],
  };
};
```

App.js
```
import React from "react";
import { Text, View } from "react-native";
import * as Analytics from "expo-firebase-analytics";
import logEvent from "./log-event";

// this causes a warning
logEvent();

// this does not cause a warning
Analytics.logEvent("testevent");

export default function App() {
  return (
    <View>
      <Text>Open up App.js to start working on your app!</Text>
    </View>
  );
}
```
log-event.js
```
import * as Analytics from "expo-firebase-analytics";

export default () => {
  Analytics.logEvent("testevent");
};
```
google-services.json + app.json are pretty standard. It actually works, when I invoke expo-firebase-analytics directly in App.js. 

 
### Steps to Reproduce

- Start a new project with expo init.
- Replace App.js with file and create log-event.js.
- Add your firebase config to app.json + associated google-services.json (maybe not needed for android client)
- Expo start. Wait about 5 seconds after seeing the <Text> and the warning will show up. 
- comment out logEvent(); and the warning goes away, even though it runs the same code as below. So long as Analytics.logEvent() is in the same file, no warning is shown.

### Expected Behavior vs Actual Behavior
I would hope to see no warning, or a warning that is a bit more clear about what request is failing.
This is what I'm getting
>  [Unhandled promise rejection: TypeError: Network request failed]
> - node_modules/whatwg-fetch/dist/fetch.umd.js:527:17 in setTimeout$argument_0
> - node_modules/react-native/Libraries/Core/Timers/JSTimers.js:130:14 in _callTimer
> - node_modules/react-native/Libraries/Core/Timers/JSTimers.js:383:16 in callTimers
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:416:4 in __callFunction
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:109:6 in __guard$argument_0
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:364:10 in __guard
> - node_modules/react-native/Libraries/BatchedBridge/MessageQueue.js:108:4 in callFunctionReturnFlushedQueue
> * [native code]:null in callFunctionReturnFlushedQueue

It seems like the call stack will include paths to unrelated code sometimes, but is an actual error in that case.

**Comment by brentvatne** [comment]
this happens in a SDK 38 app with Expo client 2.16.1 with the provided code as well, so it is unrelated to SDK 39.

**Comment by ioness** [comment]
> this happens in a SDK 38 app with Expo client 2.16.1 with the provided code as well, so it is unrelated to SDK 39.

Same here...

In my case it seems that logging navigation collides with any other simulaneous logEvent call

( Sorry for the Spam )

**Comment by IjzerenHein** [comment]
Hi! Thanks for the elaborate explanation. 👍  I was able to find out what the problem was using your help. It was related to a http header (see PR #106006) which was not set correctly and this caused a problem when 1 or more events were send as a batch.

We've just published `expo-firebase-analytics@2.5.1` which fixes the problem.

**Comment by frankcalise** [comment]
I upgraded via `npm install expo-firebase-analytics@2.5.1` (tried expo install and it would always leave me at 2.5.0) and it seems as if I can still reproduce this issue. Anyone else?

```
    "expo-firebase-analytics": "2.5.1",
    "firebase": "7.9.0",
    "react-native": "https://github.com/expo/react-native/archive/sdk-39.0.3.tar.gz",
```

**Comment by IjzerenHein** [comment]
@frankcalise Could you describe in detail what problem you are experiencing and what errors you are seeing in your console?

**Comment by frankcalise** [comment]
@IjzerenHein Yeah sorry, it seems to be that same Network request issue when switching screens rather quickly (so maybe batched events). I only can get it to happen on Android I believe.

```
[Unhandled promise rejection: TypeError: Network request
failed]
- node_modules\whatwg-fetch\dist\fetch.umd.js:527:17 in setTimeout$argument_0
- node_modules\react-native\Libraries\Core\Timers\JSTimers.js:130:14 in _callTimer
- node_modules\react-native\Libraries\Core\Timers\JSTimers.js:383:16 in callTimers
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:416:4 in __callFunction
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:109:6 in __guard$argument_0
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:364:10 in __guard
- node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:108:4 in callFunctionReturnFlushedQueue
* [native code]:null in callFunctionReturnFlushedQueue
```

```
    "expo-firebase-analytics": {
      "version": "2.5.1",
      "resolved": "https://registry.npmjs.org/expo-firebase-analytics/-/expo-firebase-analytics-2.5.1.tgz",
      "integrity": "sha512-DdETV+CuanaoKg06dy9G5eBBgpPTsuvKmMDn4+NzKgNFcXiBoTOt4IgL3RccqH/UoYkVFchumIMsAniLjAc0Cg==",
      "requires": {
        "expo-firebase-core": "~1.2.0"
      }
    },
```

**Comment by IjzerenHein** [comment]
@frankcalise Which version of whatwg-fetch do you have installed in your node_modules?

**Comment by frankcalise** [comment]
@IjzerenHein 3.4.1 

```
    "whatwg-fetch": {
      "version": "3.4.1",
      "resolved": "https://registry.npmjs.org/whatwg-fetch/-/whatwg-fetch-3.4.1.tgz",
      "integrity": "sha512-sofZVzE1wKwO+EYPbWfiwzaKovWiZXf4coEzjGP9b2GBVgQRLQUZ2QcuPpQExGDAW5GItpEm6Tl4OU5mywnAoQ=="
    },
```

**Comment by IjzerenHein** [comment]
@frankcalise Okay, that's the latest version for which it should definitely be compatible with. If you'd like to help out and do some testing. I think the problem might be with the `user-agent` header. Could you try capitalizing the user-agent header to:

```
'User-Agent': `Expo/${Constants.nativeAppVersion}`,
```

in your local `node_modules/expo-firebase-analytics/build/ExpoFirebaseAnalytics.js` (line 48)

and see whether that makes any change?
You can also try removing the header altogether.
This would greatly help in identifying where the problem is. Thanks!

**Comment by frankcalise** [comment]
@IjzerenHein I tried both of those changes. I still received the same warning. Let me know how else I can help.

To confirm this was related, I went and removed the screen change logging event.

```
<AppContainer
              ref={NavigationService.setTopLevelNavigator}
              uriPrefix={uriPrefix}
              onNavigationStateChange={(prevState, currentState, action) => {
                if (Constants.isDevice) {
                  const currentRouteName = getActiveRouteName(currentState);
                  const previousRouteName = getActiveRouteName(prevState);

                 /*
                  if (previousRouteName !== currentRouteName) {
                    Analytics.setCurrentScreen(
                      currentRouteName,
                      currentRouteName
                    );
                  }
                */
                }
              }}
            />
```

and the warning goes away when testing. My steps to test were just switching tabs back and forth rapidly a couple of times. Then after a short delay the warning is logged to the console.

**Comment by IjzerenHein** [comment]
Hmm, that's unfortunate, I had hoped that would have worked. It's very strange that you get the `TypeError: Network request failed]` error, as it is quite a normal `fetch` request with no special headers when the user-agent is removed.
Are you really sure that you're loading `expo-firebase-analytics@2.5.1` or the modified code and are not accidentally using some cached version?

**Comment by frankcalise** [comment]
@IjzerenHein Hmmm ok I'll double check all that, aside from expo start -c, I guess just blow away node_modules and go with a fresh npm install? I'll give those two a shot again. Thanks.

**Comment by frankcalise** [comment]
It seems good now, I deleted node_modules and package-lock.json. I'm hosting from my mac this time which is different than my windows box (traveling at the moment). But it seems to be OK. I only saw two other warnings but not related to the firebase package. 

This one: https://github.com/expo/expo/issues/5944
and the splash screen one but I've seen that floating around.

Thanks for your help and sorry for bringing up a closed issue.
